x-pack/filebeat/input/awscloudwatch/input.go

```diff
@@ -0,0 +1,316 @@
+			nextToken = *resp.NextToken
+		}
+
+		logEvents := resp.Events
+		in.logger.Debugf("Processing #%v events", len(logEvents))
+
+		err = in.processLogEvents(logEvents)
+		if err != nil {
+			err = errors.Wrap(err, "processLogEvents failed")
+			in.logger.Error(err)
+			cancelFn()
+		}
+
+		init = false
+
+		// FilterLogEvents has a limit of 5 transactions per second (TPS)/account/Region
+		// This sleep is to avoid hitting the API limit.
+		in.logger.Debugf("sleeping for %v before making FilterLogEvents API call again", in.config.APISleep)
+		time.Sleep(in.config.APISleep)
+		in.logger.Debug("done sleeping")
+	}
+	return nil
+}
+
+func (in *awsCloudWatchInput) constructFilterLogEventsInput(startTime int64, endTime int64, nextToken string) *cloudwatchlogs.FilterLogEventsInput {
+	filterLogEventsInput := &cloudwatchlogs.FilterLogEventsInput{
+		LogGroupName: awssdk.String(in.config.LogGroupName),
+		StartTime:    awssdk.Int64(startTime),
+		EndTime:      awssdk.Int64(endTime),
+	}
+
+	if len(in.config.LogStreams) > 0 {
+		filterLogEventsInput.LogStreamNames = in.config.LogStreams
+	}
+
+	if in.config.LogStreamPrefix != "" {
+		filterLogEventsInput.LogStreamNamePrefix = awssdk.String(in.config.LogStreamPrefix)
+	}
+
+	if nextToken != "" {
+		filterLogEventsInput.NextToken = awssdk.String(nextToken)
+	}
+	return filterLogEventsInput
+}
+
+func getStartPosition(startPosition string, currentTime time.Time, prevEndTime int64, scanFrequency time.Duration) (startTime int64, endTime int64) {
+	switch startPosition {
+	case "beginning":
+		if prevEndTime != int64(0) {
+			return prevEndTime, currentTime.UnixNano() / int64(time.Millisecond)
+		}
+		return 0, currentTime.UnixNano() / int64(time.Millisecond)
+	case "end":
+		if prevEndTime != int64(0) {
+			return prevEndTime, currentTime.UnixNano() / int64(time.Millisecond)
+		}
+		return currentTime.Add(-scanFrequency).UnixNano() / int64(time.Millisecond), currentTime.UnixNano() / int64(time.Millisecond)
+	}
+	return
+}
+
+func (in *awsCloudWatchInput) processLogEvents(logEvents []cloudwatchlogs.FilteredLogEvent) error {
+	for _, logEvent := range logEvents {
+		event := createEvent(logEvent, in.config.LogGroupName, in.config.RegionName)
+		err := in.forwardEvent(event)
+		if err != nil {
+			err = errors.Wrap(err, "forwardEvent failed")
+			in.logger.Error(err)
+			return err
+		}
+	}
+	return nil
+}
+
+func createEvent(logEvent cloudwatchlogs.FilteredLogEvent, logGroup string, regionName string) beat.Event {
+	event := beat.Event{
+		Timestamp: time.Unix(*logEvent.Timestamp/1000, 0).UTC(),
+		Fields: common.MapStr{
+			"message":        *logEvent.Message,
+			"log.file.path":  logGroup + "/" + *logEvent.LogStreamName,
+			"event.ingested": time.Now(),
+			"awscloudwatch": common.MapStr{
+				"log_group":      logGroup,
+				"log_stream":     *logEvent.LogStreamName,
+				"ingestion_time": time.Unix(*logEvent.IngestionTime/1000, 0),
+				"timestamp":      time.Unix(*logEvent.Timestamp/1000, 0),
```

Comment:
I don't think this field is needed, having we already store it under `@timestamp`